README.md
# cucumber

FIXME: description

## Installation

Download from http://example.com/FIXME.

## Usage

FIXME: explanation

    $ java -jar cucumber-0.1.0-standalone.jar [args]

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2016 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased][unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-10-26
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-10-26
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[unreleased]: https://github.com/your-name/cucumber/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/cucumber/compare/0.1.0...0.1.1


test/api_test/xml_test.clj
(ns api-test.xml-test
  (:require [clojure.test :refer :all]
            [api-test.core :refer :all]
            [api-test.mq-send :refer :all]
            [api-test.xml :refer :all]))

(def test-response "<?xml version='1.0'?><response><NXHeader><Version></Version><ServiceName>AMD.ENSEMBLE.CREATE_BAN</ServiceName><ApplRef>1477923078_BOOST_BAN</ApplRef><ReplyCompCode>0</ReplyCompCode></NXHeader><header></header><createBan><CreateBanRespInfo><banId>754776797</banId></CreateBanRespInfo></createBan></response>")

(def test-values (lazy-seq '({:content "888888888", :tag "createBan>CreateBanRespInfo>banId"} {:content "1142664520", :tag "NXHeader>ReplyCompCode"})))

(def bad-test-values (lazy-seq '({:content "888888888", :tag "CreateBan>CreateBanRespInfo>banId"} {:content "1142664520", :tag "NXHeader>ReplyCompCode"})))

(def two-bad-test-values (lazy-seq '({:content "888888888", :tag "CreateBan>CreateBanRespInfo>banId"} {:content "1142664520", :tag "NXHeader>rReplyCompCode"})))


(deftest get-header
  (testing "getting a value out of the NXHeader"
    (is (= "0" (get-value "NXHeader>ReplyCompCode" test-response)))))

(deftest get-value-1
  (testing "getting a value"
    (is (= "AMD.ENSEMBLE.CREATE_BAN"  (get-value "NXHeader>ServiceName" test-response)))))

(deftest get-value-2
  (testing "getting a value"
    (is (= "754776797"  (get-value "createBan>CreateBanRespInfo>banId" test-response)))))

(deftest get-value-3
  (testing "getting a value thats not there"
    (is (not= "754776797"  (get-value "CreateBan>CreateBanRespInfo>banId" test-response)))))

(deftest set-value
  (testing "setting a value in a message"
    (let [message (set-values test-values test-response)]
      (is (and
           (= "888888888" (get-value "createBan>CreateBanRespInfo>banId" message))
           (= "1142664520" (get-value "NXHeader>ReplyCompCode" message)))))))

(deftest set-value-1
  (testing "setting a value in a message with one bad line"
    (let [message (set-values bad-test-values test-response)]
      (is (and
           (not= "888888888" (get-value "createBan>CreateBanRespInfo>banId" message))
           (= "1142664520" (get-value "NXHeader>ReplyCompCode" message)))))))


(deftest set-value-2
  (testing "setting a value in a message with two bad lines"
    (let [message (set-values two-bad-test-values test-response)]
      (is (and
           (not= "888888888" (get-value "createBan>CreateBanRespInfo>banId" message))
           (not= "1142664520" (get-value "NXHeader>ReplyCompCode" message)))))))



test/api_test/core_test.clj
(ns api-test.core-test
  (:require [clojure.test :refer :all]
            [api-test.core :refer :all]
            [api-test.mq-send :refer :all]
            [api-test.xml :refer :all]))

(deftest test-store-values
  (testing "store and get values"
   (update-stored-value {:$BANID "999"})
   (is (= "999" (get-stored-value :$BANID)))))

(deftest run-cukes
  ;(. cucumber.api.cli.Main (main (into-array ["--plugin" "pretty" "--glue" "/Users/fhitchen/git/api-test/api-test/test/features/step_definitions" "/Users/fhitchen/git/api-test/api-test/test/features"])))
  )


test/features/step_definitions/basic_stepdefs.clj
(use 'api-test.core)
(use 'clojure.test)
(use 'api-test.mq-send)
(use 'api-test.xml)

(Given #"^message update-subscriber\.xml$" []
  (comment  Write code here that turns the phrase above into concrete actions  )
  (throw (cucumber.api.PendingException.)))

(Given #"^the message \"([^\"]*)\" with the following values:$" [message-file data]
       (reset! x-values (table->rows data))
       ;(println (pr-str "x-values= " (table->rows data)))
       (reset! message (set-values (table->rows data) (slurp (str "resources/messages/" message-file)))))

(Given #"^we send the message$" []
       (mq-send @message))

(Given #"^we send the message with the following modified values:$" [data]
       (let [m (set-values (table->rows data) @message)]
         (mq-send m)))

(When #"^we receive the response$" []
      (reset! response (mq-receive))
      (assert (not= nil @response)))

(Then #"^the message header response value named \"([^\"]*)\" must equal \"([^\"]*)\"$" [arg1 arg2]
      (let [value (get-value arg1 (.getText @response))]
        (assert (= (str arg2) value))))

(Then #"^the message body response value named \"([^\"]*)\" should be (\d+) characters long$" [arg1 length]
      (let [value (get-value arg1 (.getText @response))]
        (assert (= (Integer/parseInt length) (count value)))))

(Then #"^the response message tag named \"([^\"]*)\" contains the text \"([^\"]*)\"$" [tag value]
      (assert (not= nil (re-find (re-pattern value) (get-value tag (.getText @response))))))

(Then #"^store the \"([^\"]*)\" value in variable \"([^\"]*)\"$" [tag variable-name]
      (update-stored-value {(keyword variable-name) (get-value tag (.getText @response))}))

(Given #"^message add-subscriber\.xml$" []
  (comment  Write code here that turns the phrase above into concrete actions  )
  (throw (cucumber.api.PendingException.)))


src/api_test/core.clj
(ns api-test.core
  (:gen-class))

(def x-values (atom []))
(def response (atom ""))
(def stored-values (atom (hash-map)))
(def message (atom ""))


(defn update-stored-value
  [value]
  (reset! stored-values (conj @stored-values value)))
(defn get-stored-value
  [value]
  (value @stored-values))


;(def test-values (lazy-seq '({:content "$BANID", :tag "createBan>createBanRespInfo>banId"} {:content "1142664520", :tag "NXHeader>rReplyCompCode"})))

(defn replace-variables [values]
  (map #(if-let [v (re-find  #"^\$.*" (:content %))]
          (do
            (println (str "found var " (:content %)))
            (println (str "replace with " %))
            (assoc-in % [:content] (get-stored-value (keyword v) )))
          %) values))

(class (update-stored-value {:$BANID "nine" }))
;(get-stored-value :$BANID)
;(replace-variables test-values)

;(assoc-in {:content "$BANID" :tag "xxx" } [:content] "9999")

(defn -main
  "I don't do a whole lot ... yet."
  [& args]
  (println "Hello, World!"))






src/api_test/mq_send.clj
(ns api-test.mq-send)
;; Simple example of sending a message to Websphere MQ with Clojure
;;
;; Uses the Websphere MQ JMS classes (i.e. JMS-like in API, but
;; submits to a MQ queue)
;; 
;; Author:   Frederico Munoz <frederico.munoz@pt.ibm.com>
;; Date:     20-Jun-2013
;; Keywords: mq, websphere, messaging, jms
;;
;; Copying and distribution of this file, with or without modification,
;; are permitted in any medium without royalty provided the copyright
;; notice and this notice are preserved.  This file is offered as-is,
;; without any warranty.
;;
;; See http://www.ibm.com/developerworks/downloads/ws/wmq/ for all the
;; client downloads available; the following jar files should be in
;; the classpath: 
;; - com.ibm.mqjms.jar 
;; - jms.jar
;;
;; This is just a quick example that tries to keep it as simple as
;; possible; this means no error checking, catch/try, or even any
;; degree of optimisation that could make the example less obvious, no
;; checking return codes.
;;
;; Use of leiningen is recommended for any more advanced use, but
;; assuming all the jars are in the current directory the following
;; should work
;;
;; $ java -cp clojure.jar:com.ibm.mqjms.jar:jms.jar clojure.main mqm.clj
;; Sat Jun 22 17:53:49 WEST 2013 : Sent 
;;   JMSMessage class: jms_text
;;   JMSType:          null
;;   JMSDeliveryMode:  2
;;   JMSExpiration:    0
;;   JMSPriority:      4
;;   JMSMessageID:     ID:414d5120494f432e4d422e514d202020ad7cc45102c30f20
;;   JMSTimestamp:     1371920029641
;;   JMSCorrelationID: null
;;   JMSDestination:   queue:///PT.TEST.IN
;;   JMSReplyTo:       null
;;   JMSRedelivered:   false
;;     JMSXAppID: WebSphere MQ Client for Java
;;     JMSXDeliveryCount: 0
;;     JMSXUserID: mqmconn     
;;     JMS_IBM_PutApplType: 28
;;     JMS_IBM_PutDate: 20130622
;;     JMS_IBM_PutTime: 16534966
;; My soul is like a shepherd. It knows wind and sun, walking hand in hand with the Seasons
;; Delivered to queue:///PT.TEST.IN
;; $

(import 'com.ibm.jms.JMSMessage)
(import 'com.ibm.jms.JMSTextMessage)
(import 'com.ibm.mq.jms.JMSC)
(import 'com.ibm.mq.jms.MQQueue)
(import 'com.ibm.mq.jms.MQQueueConnection)
(import 'com.ibm.mq.jms.MQQueueConnectionFactory)
(import 'com.ibm.mq.jms.MQQueueReceiver)
(import 'com.ibm.mq.jms.MQQueueSender)
(import 'com.ibm.mq.jms.MQQueueSession)
(import 'javax.jms.Session)


;; Main variables, most likely candidates for additional command line arguments
;; Change them according to your environment (the message can be supplied in the cli)
(def mq-qm "QM1")             ; MQ queue manager name
(def mq-channel "SYSTEM.ADMIN.SVRCONN")  ; MQ channel name
(def mq-port 32768)                  ; MQ port
(def mq-queue "REQUEST_QUEUE")       ; MQ queue name
(def mq-host "cmildev136")          ; MQ server
(def mq-message (if (seq (first *command-line-args*)) ; Message, either the first 
                  (first *command-line-args*)         ; command line argument of a snippet from Pessoa's
                  "My soul is like a shepherd. It knows wind and sun, walking hand in hand with the Seasons"))

(defn- mq-connection-factory
  "Sets and returns a CONNECTION-FACTORY using the provided options"
  [{:keys [host port transport qm channel]}]
  (let [connection-factory (MQQueueConnectionFactory.)]
    (doto connection-factory
      (.setHostName host)
      (.setPort port)
      (.setTransportType transport)
      (.setQueueManager qm)
      (.setChannel channel))
    connection-factory))

(defn- mq-send-message [connection-factory queue message]
  "Sends MESSAGE, a string, to the QUEUE using CONNECTION-FACTORY"
  (let [conn (.createQueueConnection connection-factory)
        session (.createQueueSession conn false Session/AUTO_ACKNOWLEDGE)
        queue (.createQueue session (str "queue:///" queue))
        sender (.createSender session queue)
        message (.createTextMessage session message)]
    (.start conn)
    (.send sender message)
    (.close sender)
    (.close conn)
    ;(println (str (java.util.Date.) " : Sent message follows" message
    ;"\nDelivered to " queue))
    ))

(defn- mq-receive-message [connection-factory queue]
  "Sends MESSAGE, a string, to the QUEUE using CONNECTION-FACTORY"
  (let [conn (.createQueueConnection connection-factory)
        session (.createQueueSession conn false Session/AUTO_ACKNOWLEDGE)
        queue (.createQueue session (str "queue:///" queue))
        receiver (.createReceiver session queue) 
        message (.createTextMessage session "dummy")]
    (.start conn)
    (let [message (.receive receiver 5000)]
      (.close receiver)
      (.close conn)
      ;(println (str (java.util.Date.) " : Reveived message " message "\nfrom " queue))
      message)))

(defn mq-send
  "Function to send message"
  [message & args]
  (let [mq-connection (mq-connection-factory {:host mq-host
                                              :port mq-port
                                              ;:transport JMSC/MQJMS_TP_CLIENT_MQ_TCPIP
                                              :transport JMSC/MQJMS_TP_CLIENT_MQ_TCPIP
                                              :qm mq-qm
                                              :channel mq-channel})]
    (mq-send-message mq-connection mq-queue message)))

(defn mq-receive
  "Function to receive message"
  [& args]
  (let [mq-connection (mq-connection-factory {:host mq-host
                                              :port mq-port
                                              ;:transport JMSC/MQJMS_TP_CLIENT_MQ_TCPIP
                                              :transport JMSC/MQJMS_TP_CLIENT_MQ_TCPIP
                                              :qm mq-qm
                                              :channel mq-channel})]
    (mq-receive-message mq-connection "RESPONSE_QUEUE")))


;; Just send it already
;(mq-example)
;(mq-receive)








;;;; End of file


src/api_test/xml.clj
(ns api-test.xml
  (:require
   [clojure.zip :as zip]
   [clojure.data.zip :as zf]
   [clojure.data.xml :as xml]
   [clojure.data.zip.xml :as zx]
   [clojure.string :as str]
   [clojure.pprint :as p])
  (:import [java.io ByteArrayInputStream]))

(defn parse [string]
  (xml/parse
   (ByteArrayInputStream.
    (.getBytes (.trim string)))))

(defn uuid [] (str (java.util.UUID/randomUUID)))

;(def test-response "<?xml version='1.0'?><response><NXHeader><Version></Version><ServiceName>AMD.ENSEMBLE.CREATE_BAN</ServiceName><ApplRef>1477923078_BOOST_BAN</ApplRef><ReplyCompCode>0</ReplyCompCode></NXHeader><header></header><createBan><CreateBanRespInfo><banId>754776797</banId></CreateBanRespInfo></createBan></response>")

(;def test-values (lazy-seq '({:content "888888888", :tag
 ;"createBan>createBanRespInfo>banId"} {:content "1142664520", :tag
 ;"NXHeader>rReplyCompCode"}))
 )

;(println (pr-str test-values))


(defn set-values
  "given a lazy sequence of cucumber table values of :tag path and :content reset the :content of the matching paths in the message"
  [values message]
  (loop [zipper (zip/xml-zip (parse message)) line values]
    (if (= () line)
      (xml/indent-str (zip/root zipper))
      (do
        (let [paths (str/split (:tag (first line)) #">")
              matches (map #(zx/tag= (keyword %)) paths)
              edit { :tag (keyword (last paths)) :content (:content (first line)) }
              new-loc (apply zx/xml1-> zipper matches)]
          (let [loc (if (not= nil new-loc)
                      (zip/root (zip/replace new-loc edit))
                      (zip/root zipper))]
            (recur (zip/xml-zip loc) (rest line))))))))


;(println (xset-values test-values test-response))

(defn get-value
  [path message]
  (let [paths (str/split path #">")
        matches (map #(zx/tag= (keyword %)) paths)
        zipper (zip/xml-zip (parse message))]
    (apply zx/xml1-> zipper (conj (into [] matches) zx/text))))

(defn print-values
  "print set values out of input message"
  [message-file]
  (let [message (slurp (str "resources/messages/" message-file))
        zipper (zip/xml-zip (parse message))]
    (loop [z zipper]
      (if (zip/end? z)
        true
        (do
          (if (string? (zip/node z))
            (do
              (loop [zz z path (str " \"" (zip/node z) "\"")]
                (if-not (zip/up zz)
                  (println (str/replace path #"request>(.*)> (\".*\")" "\t| \"$1\" | $2 |"))
                  (do
                    (recur (zip/up zz) (str (name (:tag (zip/node (zip/up zz)))) ">" path)))))))
          (recur (zip/next z)))))))


;(print-values "bottom.xml")











